### PR TITLE
Optimize CatalogRecycleBin query table with same name

### LINKED ISSUE #3585: Failed to acquire catalog lock when drop the same table name
### Steps to reproduce the behavior (Required)
1. FE have a lot of table
2. drop same table name

### Expected behavior (Required)
delete success
### Real behavior (Required)
```
com.starrocks.common.DdlException: Failed to acquire catalog lock. Try again
        at com.starrocks.catalog.Catalog.dropDb(Catalog.java:2701) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.DdlExecutor.execute(DdlExecutor.java:106) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.StmtExecutor.handleDdlStmt(StmtExecutor.java:912) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.StmtExecutor.execute(StmtExecutor.java:427) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.ConnectProcessor.handleQuery(ConnectProcessor.java:268) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.ConnectProcessor.dispatch(ConnectProcessor.java:415) ~[starrocks-fe.jar:?]
        at com.starrocks.qe.ConnectProcessor.processOnce(ConnectProcessor.java:651) ~[starrocks-fe.jar:?]
        at com.starrocks.mysql.nio.ReadListener.lambda$handleEvent$0(ReadListener.java:55) ~[starrocks-fe.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_252]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_252]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_252]
2022-02-21 17:54:28,714 WARN (starrocks-mysql-nio-pool-7|363) [Catalog.tryLock():766] catalog lock is held by: dump thread: starrocks-mysql-nio-pool-151, id: 1272
    com.starrocks.catalog.CatalogRecycleBin.eraseTableWithSameNama(CatalogRecycleBin.java:310)
    com.starrocks.catalog.CatalogRecycleBin.recycleTable(CatalogRecycleBin.java:112)
    com.starrocks.catalog.Catalog.unprotectDropTable(Catalog.java:4847)
    com.starrocks.catalog.Catalog.unprotectDropDb(Catalog.java:2788)
    com.starrocks.catalog.Catalog.dropDb(Catalog.java:2762)
    com.starrocks.qe.DdlExecutor.execute(DdlExecutor.java:106)
    com.starrocks.qe.StmtExecutor.handleDdlStmt(StmtExecutor.java:912)
    com.starrocks.qe.StmtExecutor.execute(StmtExecutor.java:427)
    com.starrocks.qe.ConnectProcessor.handleQuery(ConnectProcessor.java:268)
    com.starrocks.qe.ConnectProcessor.dispatch(ConnectProcessor.java:415)
    com.starrocks.qe.ConnectProcessor.processOnce(ConnectProcessor.java:651)
    com.starrocks.mysql.nio.ReadListener.lambda$handleEvent$0(ReadListener.java:55)
    com.starrocks.mysql.nio.ReadListener$$Lambda$308/271230035.run(Unknown Source)
    java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    java.lang.Thread.run(Thread.java:748)
```
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #3585

## Problem Summary(Required) ：

Reconstruct CatalogRecycleBin into triple DbId, TableId, Info.
Accelerates the query to adapt to a large number of new and deleted many many of tables scenarios.

### CODE REVIEW

**Comment by Astralidea** [comment]
run all

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 1]
```diff
@@ -256,40 +258,42 @@ public synchronized void replayEraseDatabase(long dbId) {
...
-        } // end for tables
-        if (!tableIdList.isEmpty()) {
-            Catalog.getCurrentCatalog().getEditLog().logEraseMultiTables(tableIdList);
-            LOG.info("multi erase write log finished. erased {} table(s)", currentEraseOpCnt);
         }
     }
 
     private synchronized void eraseTableWithSameName(long dbId, String tableName) {
-        Iterator<Map.Entry<Long, RecycleTableInfo>> iterator = idToTable.entrySet().iterator();
+        Iterator<Map.Entry<Long, RecycleTableInfo>> iterator = idToTable.row(dbId).entrySet().iterator();
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:296]: Why does the function will hold the lock too long?

Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:296]: I think it's caused by too many elements in idToTable, but I can't think of other reasons, and it doesn't reproduce locally.

Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:296]: Database and partition also have the same problem.

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 2]
```diff
@@ -114,21 +119,22 @@ public synchronized boolean recycleTable(long dbId, Table table) {
...
+        RecycleTableInfo tableInfo = idToTableInfo.row(dbId).get(tableId);
         if (tableInfo != null) {
             return tableInfo.getTable();
         }
         return null;
     }
 
     public synchronized List<Table> getTables(long dbId) {
-        return idToTable.values().stream()
+        return idToTableInfo.values().stream()
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:137]: Better to use idToTableInfo.row(dbId) to get the tables.

File: fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java [hunk 3]
```diff
@@ -256,69 +261,74 @@ public synchronized void replayEraseDatabase(long dbId) {
...
+        RecycleTableInfo tableInfo = nameToTableInfoDBLevel.get(tableName);
+        if (tableInfo == null) {
+            return;
+        }
+        Table table = tableInfo.getTable();
 
-                iterator.remove();
-                idToRecycleTime.remove(table.getId());
-                LOG.info("erase table[{}-{}], because table with the same name is recycled", table.getId(), tableName);
+        if (table.getName().equals(tableName)) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:307]: This check is not needed.

**Review by wyb — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by chaoyli** [comment]
run starrocks_fe_unittest_coverage

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 31 / 89 (34.83%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/CatalogRecycleBin.java|30|88|34.09%|
|:large_blue_circle:|com/starrocks/catalog/Catalog.java|1|1|100.00%|

**Pull request merged**